Lib/pathlib/_abc.py

```diff
@@ -87,6 +87,128 @@ def isabs(self, path):
+        # would be a single network share mounted in one location via NFS, and
+        # in another location via CIFS), so we simply checks whether the
+        # other path is lexically equal to, or within, this path.
+        if source == target:
+            err = OSError(EINVAL, "Source and target are the same path")
+        elif source in target.parents:
+            err = OSError(EINVAL, "Source path is a parent of target path")
+        else:
+            return
+        err.filename = str(source)
+        err.filename2 = str(target)
+        raise err
+
+    def copy(self, source, target):
+        """Copy the given file or directory tree to the given target."""
+        self.ensure_distinct_paths(source, target)
+        if self.preserve_metadata:
+            metadata_keys = source._readable_metadata & target._writable_metadata
+        else:
+            metadata_keys = None
+        if not self.follow_symlinks and source.is_symlink():
+            self.copy_symlink(source, target, metadata_keys)
+        elif source.is_dir():
+            self.copy_dir(source, target, metadata_keys)
+        else:
+            self.copy_file(source, target, metadata_keys)
+
+    def copy_dir(self, source, target, metadata_keys, dir_entry=None):
+        """Copy the given directory to the given target."""
+        if metadata_keys:
+            metadata = source._read_metadata(metadata_keys, dir_entry=dir_entry)
+        else:
+            metadata = None
+        with source.scandir() as entries:
+            target.mkdir(exist_ok=self.dirs_exist_ok)
+            for entry in entries:
+                src = source._join_dir_entry(entry)
+                dst = target.joinpath(entry.name)
+                if not self.follow_symlinks and entry.is_symlink():
+                    self.copy_symlink(src, dst, metadata_keys, entry)
+                elif entry.is_dir():
+                    self.copy_dir(src, dst, metadata_keys, entry)
+                else:
+                    self.copy_file(src, dst, metadata_keys, entry)
+            if metadata:
+                target._write_metadata(metadata)
+
+    def copy_file(self, source, target, metadata_keys, dir_entry=None):
+        """Copy the given file to the given target."""
+        self.ensure_different_files(source, target, dir_entry)
+        if metadata_keys:
+            metadata = source._read_metadata(metadata_keys, dir_entry=dir_entry)
+        else:
+            metadata = None
+        with source.open('rb') as source_f:
+            try:
+                with target.open('wb') as target_f:
+                    copyfileobj(source_f, target_f)
+            except IsADirectoryError as e:
+                if not target.exists():
+                    # Raise a less confusing exception.
+                    raise FileNotFoundError(
+                        f'Directory does not exist: {target}') from e
+                raise
```

Comment:
Sanity checking -- does this directory error handling need to exist in `copy_file`? Can ensuring that source and target are files be handled as a precondition? No worries if 'yes', this just surprised me!